### ISSUE TITLE
When should `NIOHTTPClientUpgradeHandler` completion handler be called?

### ISSUE THREAD

**Issue by fumoboy007** [apple/swift-nio#2384]
### Expected behavior
I expected `NIOHTTPClientUpgradeHandler` [completion handler](https://github.com/apple/swift-nio/blob/5f5fa9a2b2c04934e7a1689827b590e6054a6361/Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift#L116) to be called after all upgrade operations have been complete.

### Actual behavior
The completion handler seems to be [called](https://github.com/apple/swift-nio/blob/5f5fa9a2b2c04934e7a1689827b590e6054a6361/Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift#L315) prematurely?

### If possible, minimal yet complete reproducer code (or URL to code)

https://github.com/apple/swift-nio/blob/5f5fa9a2b2c04934e7a1689827b590e6054a6361/Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift#L315

### SwiftNIO version/commit hash

5f5fa9a2b2c04934e7a1689827b590e6054a6361

**Comment by Lukasa** [comment]
I agree that this seems to be called prematurely. Unfortunately, I don't think we can easily change that without risking a break to running programs that rely on that behaviour.

We should certainly update the documentation to call out when this closure is _actually_ called, and we should consider adding a new closure that is called when the upgrade actually has completed.
